### ISSUE TITLE
How can we reduce type parameters further?

### ISSUE THREAD

**Issue by tomjaguarpaw** [Mikolaj/horde-ad#34]
I think we need @goldfirere to help with this.

To simplify the surface API we'd like to reduce the length number of type parameters floating around. Currently we have two (typically named `s` and `dual`).  We'd like to reduce it to one. In principle this is easy: put the two into a type-level pair! In practice this doesn't work, and I don't know why.

Have a look at 7ba5614. I package `s` and `dual` into `'(s, dual)` and implement type families to take them out again. This seems to do the trick *except* I also need the constraint `s_dual ~ '(s, dual)`.  It seems that the type checker doesn't know that every type of kind `(A, B)` is of the form `'(s, dual)`. Perhaps it is correct! After all, stuck type families don't have that form.

@goldfirere Is there some ergonomic trick we can apply to force the type checker to "pattern match" on `s_dual`?

**Comment by Mikolaj** [comment]
Let me just comment that the two parameters are not a show-stopper, because the `dual` parameter is always a type variable in user code defining neural networks. The type is concrete only in special library functions such as optimizers and, while the user may want to write or fine-tune an optimizer, the amount of code involved is negligible.

OTOH, Tom's formulation with the separate `s` (the primal component) and `dual` (the type function that creates the dual component from the primal) type parameters turns out to be much better in many respects than mine. I'm just going to steal this design choice and see it suffices to remove all warts in my old implementation or else if more of the Tom's design choices are involved. So, it's possible we will quickly want to rewrite all the bulky test code to this new presentation and knowing if we are stuck with two parameters or if we can get away with one would ensure we won't have to repeat the exercise soon after.

**Comment by tomjaguarpaw** [comment]
I'm even beginning to doubt that we want to go this direction, given that https://github.com/Mikolaj/horde-ad/commit/50181cb091f6891064e5648140223b214f6d5d65 seems successful.

Still, for my own edification I would be interested to know Richard's answer about whether one can force the type checker to "assume the structure" of a type.

**Comment by goldfirere** [comment]
> Is there some ergonomic trick we can apply to force the type checker to "pattern match" on s_dual?

No. This is a long-standing shortcoming: https://gitlab.haskell.org/ghc/ghc/-/issues/7259

There are various tricks to work around the shortcoming, essentially by assuming `tuple ~ '(Fst tuple, Snd tuple)` in a bunch of places. But it's certainly a pain to work with.

**Comment by tomjaguarpaw** [comment]
Very interesting @goldfirere , thank you.
